Change Navigation block markup on front end only

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

[Follow-up exploration](https://github.com/WordPress/gutenberg/pull/30430#issuecomment-813233147) into minimising impact of nav semantics for end users. This removes the `<ul>` from the Navigation block markup and adds it conditionally on innerBlock type, only on the server-side render, so it only applies on the front end. 

Disadvantages: markup will be different between editor and front end; ~~editor markup will be unsemantic with `<li>` elements nested directly inside the `<nav>` (we can fiddle with the Navigation link block and change its editor output too if we want to fix this).~~ UPDATE: this has now been fixed, to render all link children of the Nav as `div`s in the editor.

Advantages: users don't have to bother about adding intermediary list blocks.

~~NOTE: Spacer blocks are (unsemantically) added inside the `ul` elements for now, as I was given to understand that Spacers would be required to work inside submenus too. If this is not the case, I can remove them from the lists. Otherwise we'll need to think of another way to address the Spacer issue.~~ 

The Spacer block has been temporarily removed from submenus and will be addressed separately; #33018 and #30590 explain what is needed.

With the changes in this PR, the Navigation block should contain fully semantic markup, no matter what its contents are.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

tellthemachines committed:
Change markup on front end

jasmussen commented:
Thanks so much for trying this. This is what I see:

![nav](https://user-images.githubusercontent.com/1204802/113826341-56a30900-9782-11eb-9e11-4d432b878ecf.gif)

Outside of the little spacer complexity at the end (where the two first items should be ul wrapped, as should the last one, but not the spacer), from a pure user experience perspective, this feels like the right path forward.

I understand the reluctance, but it is such a key part of the flow that it's crucial to get right. And this is a challenge that will not be unique to the navigation block, it will come up again and again. One incoming example is the table block receiving nesting support. Basic table markup looks like this:

```
<table>
     <tr>
         <td>The table body</td>
         <td>with two columns</td>
     </tr>
</table>
```

But when you add a header or footer, the markup changes:

```
<table>
    <thead>
        <tr>
            <th colspan="2">The table header</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>The table body</td>
            <td>with two columns</td>
        </tr>
    </tbody>
</table>
```

While the nesting system could replicate that 1:1, the user experience as it exists today, with each layer seletable, it would be unwieldy. This is an aspect where our friends at TinyMCE do it so well:

![tiny tables](https://user-images.githubusercontent.com/1204802/113827468-9fa78d00-9783-11eb-8ef4-a65c85964fb2.gif)

In other words, it's a problem we need to solve. In the case of the table block, a passthrough container block as discussed in #7694 might be a key ingredient in flattening the structure and showing a single block toolbar.

In the case of the navigation block, the challenge is that the `ul` needs to show up conditionally based on the content of your block. And to enable menus that have, for example, 2 menu items, a site  logo, and then another 2 menu items, this is a challenge that needs solving. The missing UL in the editor is a tradeoff that feels worth it to me, and something I can help CSS up easily. 

@youknowriad I'd love your thoughts here, as I know you've been looking at conditional divs, serverside rendered, lately.

jasmussen commented:
Coming back to this after thinking about it more, I think the most important bit of all is that we should not put the onus on the user to figure out what type of blocks need be wrapped in what containers. Users see a navigation block and assume they can insert links to pages, perhaps search, perhaps social links. 

When framed like that, what gets output on the frontend then becomes our headache to solve, either through server side rendering like in this PR, or revisiting the structure of the markup. With the server side rendering, the trade-off of markup that diverges between editor and frontend is easily worth it, as we can maintain similar classnames to target instead of the elements directly.

CC: @noisysocks @priethor @mtias @draganescu

draganescu commented:
I still feel like having unsemantic html on the editor side should not be used as an escape hatch. 

We should use this solution only if we arrive to the conclusion that it is _impossible_ to have a good UX and semantic editor markup. For now it's just complicated to see the best solution but not impossible.

gwwar commented on packages/block-library/src/navigation/index.php:
```diff
@@ -135,7 +135,16 @@ function render_block_core_navigation( $attributes, $content, $block ) {
+		if ($inner_block->name !== "core/navigation-link" && $inner_block->name !== "core/spacer" && $is_list_open === true) {
+			$is_list_open = false;
+			$inner_blocks_html .= '</ul>';
+		}
```
Nice, much closer to expected html markup. If we go with this approach, children of ul, still need to be wrapped in an li. (For example the spacer div).

<img width="1080" alt="Screen Shot 2021-04-26 at 8 59 14 AM" src="https://user-images.githubusercontent.com/1270189/116113852-bdbf2980-a66d-11eb-9a65-debb5c9b8559.png">

gwwar reviewed (commented):
Great exploration @tellthemachines ! ✨ 

My overall impression is that we do want the overall behavior of this branch (in terms of user interactions), but two things might make us hesitate on this implementation approach:

* There's interest in solving this more generally for other wrapper blocks, as noted in the conversations from https://github.com/WordPress/gutenberg/issues/7694.
* This would be one of the first areas where we'd allow markup to intentionally diverge on the editor and may lead to some a11y usability issues

This definitely works, but if we end up choosing this approach, it'd be more of a necessary one-off since it's specific to the navigation block. We'd likely end up revisiting the problem for other wrapper blocks. 

For next steps, I suspect we'll need to explore https://github.com/WordPress/gutenberg/pull/30430, and if we hit a really bad blocker, we can double back to this approach

tellthemachines committed:
Fix rendering issue with responsive wrapper

tellthemachines commented:
The discussion in #31951 points to this solution as the best way forward, so I'm marking this PR ready for review now.

I've also fixed some conflicts and updated this PR to work with the responsive wrapper.

tellthemachines committed:
Use divs in Nav link; remove Spacer from lists

tellthemachines commented:
Updated this PR to also remove the `li` and `ul`s from Navigation Link block on the editor. Front end output remains unchanged. 

I've also excluded the Spacer block from the `ul`s, and temporarily removed it from the list of allowed children of Navigation Link as we cannot semantically include it until we make changes to how Spacer renders. I'll be experimenting with how to solve that problem in the next few days, but it should be a separate PR.

Reviews and feedback appreciated!

tellthemachines committed:
lint php

jasmussen commented:
Took it for a spin:

![nav](https://user-images.githubusercontent.com/1204802/123216000-5e449500-d4c9-11eb-8d78-2a8ce86395ae.gif)

As far as I can tell, this one threads the needle! Very impressive work 👌

gwwar commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -48,7 +48,7 @@ import { store as coreStore } from '@wordpress/core-data';
 import { name } from './block.json';
 
-const ALLOWED_BLOCKS = [ 'core/navigation-link', 'core/spacer' ];
+const ALLOWED_BLOCKS = [ 'core/navigation-link' ];
```
Did we mean to remove the spacer from being insertable from submenus? We could maybe leave it for a follow up, it might be possible to do something like modify padding on the items instead of inserting an element for the published view.

<img width="1395" alt="Screen Shot 2021-06-25 at 10 46 07 AM" src="https://user-images.githubusercontent.com/1270189/123466112-cd8eb780-d5a3-11eb-8e0b-0a4399fb7ed1.png">

As a note for the other reviewers, existing menus with a spacer still appear, but new menus won't allow spacers to be inserted in sub-menus.

tellthemachines replied:
I removed it temporarily, as there are a couple of things we need to fix for it to work properly in submenus:
* the markup issue (ideally by getting the spacer to not render markup at all, or else by wrapping it in an `li`)
* making the spacer vertical within the submenu of a horizontal nav (right now it only works vertically within vertical nav blocks)

tellthemachines replied:
Update: I created #33018 to discuss the best way forward for the Spacer markup issue, and there is #30590 for its behaviour in horizontal Nav submenus. I think we can fix these separately from this PR.

jasmussen replied:
At some "phase 2" of the block, we'll want to tackle mega-menus, where the contents of a dropdown are virtually as flexible as what you can insert in a group block, so it's something we should come back to. But definitely fine to wrap a solid V1 first.

gwwar commented:
Overall this looks pretty technically sound to me, and the user experience here feels good! I smoke tested a bit, and couldn't find any issues. 

In terms of moving this forward:
- [ ] Let's double check that folks were okay with the editor/published view . It feels that we are, I just wanted us to mark that down as a comment on the PR.
- [ ] 🔍 if there are any other styling changes needed.

draganescu commented:
@gwwar what does this mean:

> Let's double check that folks were okay with the editor/published view .

gwwar commented:
> @gwwar what does this mean:

That we were okay with diverging editor markup from published markup. This is mostly to document for other 👀 since I don't expect folks to be aware of all the conversation threads here on it.

tellthemachines committed:
Update e2e test selectors

tellthemachines committed:
Update Home Link semantics

draganescu reviewed (approved):
This looks good, let's :shipit:
